Implement sort function

FikriMilano opened the pull request:
Fixes #43 

Sort exists in spec https://build.fhir.org/ig/HL7/FHIRPath/#sortkeyselector-expression-asc--desc--keyselector-expression-asc--desc---collection

Implements the FHIRPath sort() function with support for ascending/descending order and multi-key sorting.

1. sort() with no arguments - Sorts collection by items themselves in ascending order
`(3 | 2 | 1).sort()  // returns (1 | 2 | 3)`
2. sort(key1, -key2, ...) with key selectors - Sorts by one or more key expressions
`Patient.name.sort(-family, -given.first())  // descending by family, then descending by first given name`

- Descending order: Prefix key selector with - (polarity operator)
- Empty values always first: Per FHIRPath spec, empty/null values sort before non-empty regardless of direction
- Multi-key sorting: Processes selectors sequentially, using subsequent keys only when previous keys are equal

FikriMilano committed:
Implement sort

FikriMilano committed:
Remove type param

FikriMilano commented on fhir-path/src/commonMain/kotlin/com/google/fhir/fhirpath/FhirPathEvaluator.kt:
```diff
@@ -436,6 +450,59 @@ internal class FhirPathEvaluator(initialContext: Any?) : fhirpathBaseVisitor<Col
+    (this as fhirpathParser.PolarityExpressionContext).expression()
+
+  /** Evaluates a block with the given item pushed onto thisStack. */
+  private fun <T> evaluateWithThis(item: Any, block: () -> T): T {
```
removed, not really

FikriMilano commented on fhir-path/src/commonMain/kotlin/com/google/fhir/fhirpath/FhirPathEvaluator.kt:
```diff
@@ -436,6 +450,59 @@ internal class FhirPathEvaluator(initialContext: Any?) : fhirpathBaseVisitor<Col
+    this is fhirpathParser.PolarityExpressionContext && getChild(0)?.text == "-"
+
+  /** Unwraps the inner expression from a polarity expression (-expr -> expr). */
+  private fun fhirpathParser.ExpressionContext.unwrapPolarity(): fhirpathParser.ExpressionContext =
```
So, again it's about inconsistency, between the FHIRPath grammar specification and the official test suite.

On the official doc and grammar file, yes we're using asc and desc.
https://build.fhir.org/ig/HL7/FHIRPath/
```
sort([keySelector: expression [asc | desc] [, keySelector: expression [asc | desc], …]]) : collection
```

https://build.fhir.org/ig/HL7/FHIRPath/grammar.html
```
sortArgument
        : expression ('asc' | 'desc')?                          #sortDirectionArgument
        ;
```

But, on our test file (tests-fhir-r4.xml), the sort test cases are using - for desc, and not using - for asc.

The PolarityExpressionContext check is admittedly a bit hacky, but it's the correct approach given that the tests expect -expression to mean descending.

Should we use asc/desc or +/- ?

FikriMilano replied:
I've written this ticket in fhir-test-cases repo https://github.com/FHIR/fhir-test-cases/issues/253

jingtang10 replied:
we should stick to the spec. actually i think for the first 2 test cases `-$this` should still work because it's just an expression to inverse the numeric value. but for the 3rd test case it should fail.

jingtang10 replied:
So I think the numeric test cases should still pass - right? can you please check?

You can leave the string test cases excluded and link to the PR in the fhir test cases repo.

FikriMilano replied:
I see, yeah.

So.

- Removed the - as descending hack. 
- For numeric cases (testSort4/5), -$this still works since it's just arithmetic negation, sorting [-3, -2, -1] ascending naturally gives descending order. 
- For strings (testSort8/10), -$this errors out as expected since you can't negate a string. Excluded those two tests.

FikriMilano commented on fhir-path/src/commonMain/kotlin/com/google/fhir/fhirpath/FhirPathEvaluator.kt:
```diff
@@ -436,6 +450,59 @@ internal class FhirPathEvaluator(initialContext: Any?) : fhirpathBaseVisitor<Col
+
+  /** Checks if this expression is a descending sort (prefixed with -). */
+  private fun fhirpathParser.ExpressionContext.isDescendingSort(): Boolean =
+    this is fhirpathParser.PolarityExpressionContext && getChild(0)?.text == "-"
```
So, the doc and grammar file says we use asc/desc, BUT, the test file is using - to represents desc.

Detail for discussion
https://github.com/google/kotlin-fhirpath/pull/41#discussion_r2753257242

FikriMilano commented on fhir-path/src/commonMain/kotlin/com/google/fhir/fhirpath/FhirPathEvaluator.kt:
```diff
@@ -413,6 +413,20 @@ internal class FhirPathEvaluator(initialContext: Any?) : fhirpathBaseVisitor<Col
         println("$name $projection")
         context
       }
+      "sort" -> {
+        // See [specification](https://build.fhir.org/ig/HL7/FHIRPath/#sort).
+        if (context.size <= 1) return context.toList()
+
+        val keySelectors = functionNode.paramList()?.expression() ?: emptyList()
+
+        if (keySelectors.isEmpty()) {
+          // sort() with no arguments: sort by $this ascending
+          context.sortedWith { a, b -> compare(a.toFhirPathType(), b.toFhirPathType()) ?: 0 }
+        } else {
+          // sort(key1, -key2, ...) with one or more key selectors
+          context.sortedWith { a, b -> compareByKeySelectors(a, b, keySelectors) }
+        }
```
If no arguments in sort(), directly compare, to skip the overhead of:
- Pushing/popping items onto thisStack
- Visiting expression nodes via evaluateWithThis()

Only works for primitive

FikriMilano commented on fhir-path-core/src/commonMain/kotlin/dev/ohs/fhir/fhirpath/FhirPathEvaluator.kt:
```diff
@@ -413,6 +413,20 @@ internal class FhirPathEvaluator(initialContext: Any?) : fhirpathBaseVisitor<Col
+        // See [specification](https://build.fhir.org/ig/HL7/FHIRPath/#sort).
+        if (context.size <= 1) return context.toList()
+
+        val keySelectors = functionNode.paramList()?.expression() ?: emptyList()
```
I tried but it gives compilation error.
`paramList()?.expression()` returns `List<fhirpathParser.ExpressionContext>`. 
`listOf({ it })` as the fallback would introduce a Kotlin lambda, which is a different type.

The idea of the if block below is to skip the overhead of:
- Pushing/popping items onto thisStack
- Visiting expression nodes via evaluateWithThis()

When there's no keySelectors.

But it only works with primitive though.

https://github.com/google/kotlin-fhirpath/pull/41/changes#r2767826871

FikriMilano committed:
Simplify sort: let - be natural negation instead of descending directive

 The sort implementation previously treated - prefix as a descending sort directive by detecting polarity expressions and reversing comparison order. This was incorrect —
  -$this is just arithmetic negation per the FHIRPath spec, which uses asc/desc for sort direction.

FikriMilano committed:
Merge branch 'main' of github.com:google/kotlin-fhirpath into update-sort-in-readme

FikriMilano committed:
spotless

FikriMilano committed:
add missing fhirPathTypeResolver

jingtang10 commented on README.md:
```diff
@@ -207,6 +206,8 @@ documented in the table below.
 | `testCombine3`                       | Implementation     |     |                                                        | As above.                                                                                                                                                               |
 | `testPrimitiveExtensions`            | Implementation     |     |                                                        | Function `hasValue` is not implemented.                                                                                                                                 |
+| `testSort8`                          | Specification/Test |     |                                                        | Test uses `-$this` for descending string sort, but spec uses `asc`/`desc`.                                                                                              |
+| `testSort10`                         | Specification/Test |     |                                                        | Test uses `-` prefix for descending sort, but spec uses `asc`/`desc`.                                                                                                   |
```
add links to the pr in the fhir test cases repo

FikriMilano replied:
adding the issue ticket, brian had worked on it locally but havent push a pr
https://github.com/FHIR/fhir-test-cases/issues/253

FikriMilano committed:
Add fhir-test-cases ticket to README.md

jingtang10 commented on fhir-path-core/src/commonMain/kotlin/dev/ohs/fhir/fhirpath/FhirPathEvaluator.kt:
```diff
@@ -590,6 +610,43 @@ internal class FhirPathEvaluator(
+      bKey == null -> 1 // Empty always first
+      else -> compare(aKey, bKey, fhirPathTypeResolver) ?: 0
+    }
+
```
can you put these private functions at the end of this file?

the public functions in this file follow the order in the grammar.

FikriMilano replied:
done

jingtang10 commented on fhir-path-core/src/commonMain/kotlin/dev/ohs/fhir/fhirpath/FhirPathEvaluator.kt:
```diff
@@ -590,6 +610,43 @@ internal class FhirPathEvaluator(
+   *
+   * Per FHIRPath spec, empty values are always sorted first.
+   */
+  private fun compareByKeySelectors(
```
i don't know if we can inline this? not feeling strongly either way but wondering which way is more readable.

FikriMilano replied:
I made it inline, looks cleaner

jingtang10 commented on fhir-path-core/src/commonMain/kotlin/dev/ohs/fhir/fhirpath/FhirPathEvaluator.kt:
```diff
@@ -590,6 +610,43 @@ internal class FhirPathEvaluator(
+  }
+
+  /** Compares two keys for sorting. Empty values always come first per FHIRPath spec. */
+  private fun compareKeys(aKey: Any?, bKey: Any?): Int =
```
same with this function actually.

FikriMilano replied:
same done

jingtang10 commented on README.md:
```diff
@@ -264,6 +263,8 @@ documented in the table below.
 | `testCombine3`                       | Implementation     |     |                                                        | As above.                                                                                                                                                               |
 | `testPrimitiveExtensions`            | Implementation     |     |                                                        | Function `hasValue` is not implemented.                                                                                                                                 |
+| `testSort8`                          | Specification/Test |     |                                                        | Test uses `-$this` for descending string sort, but spec uses `asc`/`desc`, https://github.com/FHIR/fhir-test-cases/issues/253.                                          |
+| `testSort10`                         | Specification/Test |     |                                                        | Test uses `-` prefix for descending sort, but spec uses `asc`/`desc`, https://github.com/FHIR/fhir-test-cases/issues/253.                                               |
```
move these two lines up so that they follow the same order as before.

FikriMilano replied:
done
